File: Chapter 4/Section 4.1/nuggets.cpp
/*
ID: hsfncd31
TASK: nuggets
LANG: C++                 
*/
#include <fstream>
#include <iostream>
#include <vector>

static const int MAX_N = 256;
bool is_possible[MAX_N * MAX_N + 1];

int gcd(int a, int b) {
    if (a < b)
        std::swap(a, b);
    for (int r; r = a % b; ) {
        a = b;
        b = r;
    }
    return b;
}

int main() {
    std::fstream in("nuggets.in", std::fstream::in), out("nuggets.out", std::fstream::out);

    int n;
    in >> n;
    std::vector<int> options;
    for (int _ = 0, x; _ < n; ++_) {
        in >> x;
        is_possible[x] = true;
        options.push_back(x);
    }

    bool all_possible = is_possible[1];
    int d = options[0];
    for (int x: options)
        d = gcd(d, x);
    bool infinite_impossible = (d != 1);
    if (all_possible || infinite_impossible) {
        out << "0\n";
        return 0;
    }

    std::vector<int> possible_numbers;
    int largest_impossible;
    for (int x = 1; ; ++x) {
        if (possible_numbers.size() >= (x + 1) / 2)
            break;
        if (is_possible[x]) {
            possible_numbers.push_back(x);
            continue;
        }
        for (int a: possible_numbers) {
            if (is_possible[x - a]) {
                is_possible[x] = true;
                possible_numbers.push_back(x);
                break;
            }
            if (a >= x / 2)
                break;
        }
        if (!is_possible[x])
            largest_impossible = x;
    }

    out << largest_impossible << '\n';

    in.close();
    out.close();
}

File: Chapter 4/Section 4.1/nuggets.py
"""
ID: hsfncd31
LANG: PYTHON3
TASK: nuggets
"""
import os
import typing
import functools
import math
import itertools


def main():
    base_filename = 'test' if os.name == 'nt' else 'nuggets'
    with open('{}.in'.format(base_filename), 'r') as infile,\
            open('{}.out'.format(base_filename), 'w') as outfile:  # type: typing.IO[str]
        def get_line() -> str:
            return infile.readline().rstrip('\n')

        def out_print(*args, **kwargs) -> None:
            kwargs['file'] = outfile
            print(*args, **kwargs)

        n = int(get_line())
        options = { int(get_line()) for _ in range(n) }

        # just a guess
        infinite_largest = bool(functools.reduce(math.gcd, options) != 1)
        all_possible = bool(min(options) == 1)
        if infinite_largest or all_possible:
            out_print('0')
            return

        # to find impossible numbers, start from 1, flag if all possible number smaller doesn't add up

        # maybe always try to generate smallest number that is larger than the largest generated

        # For any k, k = floor(k / 2) + ceil(k / 2).
        # So if [a, 2a] are all possible,
        # we can generate [2a, 4a], [4a, 8a] etc.
        # impossible_numbers = []
        # impossible_numbers_set = set()
        # for x in itertools.count(1):
        #     if impossible_numbers and x // 2 >= impossible_numbers[-1] + 1:
        #         out_print(impossible_numbers[-1])
        #         break
        #     possible = False
        #     if x in options:
        #         possible = True
        #     else:
        #         for i in range(2, x // 2 + 1):
        #             if i not in impossible_numbers and x - i not in impossible_numbers:
        #                 possible = True
        #                 break
        #     if not possible:
        #         impossible_numbers.append(x)
        #         impossible_numbers_set.add(x)

        # At any time, if size of possible numbers is bigger than half of the next number,
        # the next number must be possible, retaining the feature, making all subsequent numbers possible.

        # Say if initial numbers are a and a + 1, then [ka, k(a + 1)] will be possible,
        # from k = a - 1, all numbers in [ka, k(a + 1)) will be possible,
        # so (a - 2)a + a - 1 will be the largest impossible number.
        max_i = 256
        is_possible = [False] * (max_i ** 2 + 1)
        possible_numbers = []
        largest_impossible = None
        for x in itertools.count(1):
            if len(possible_numbers) >= (x + 1) // 2:
                # all larger numbers are possible
                break
            if x in options:
                is_possible[x] = True
                possible_numbers.append(x)
                continue
            for possible_number in possible_numbers:
                if possible_number > x // 2:
                    break
                if is_possible[x - possible_number]:
                    is_possible[x] = True
                    possible_numbers.append(x)
                    break
            if not is_possible[x]:
                largest_impossible = x

        out_print(largest_impossible)



if __name__ == '__main__':
    main()
